CMakeLists.txt

```diff
@@ -23,6 +39,32 @@ if(NOT CMAKE_CXX_STANDARD MATCHES "^(17|20)$")
   message(FATAL_ERROR "CMAKE_CXX_STANDARD must be set to 17 or 20")
 endif()
 
+#-----------------------------------------------------------------------------
+# Check and enable gold linker options
+#-----------------------------------------------------------------------------
+# This is an adaptation of the original snippet in SimpleITK:
+# https://raw.githubusercontent.com/SimpleITK/SimpleITK/master/CMake/sitkUseGoldLinker.cmake
+
+include(CheckCXXCompilerFlag)
+
+if (MSVC)
+  set(_COMPILER_SUPPORTS_GOLD_LINKER 0)
+else()
+  set(CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS} -fuse-ld=gold")
+  check_cxx_source_compiles("int main() {return 0;}" _COMPILER_SUPPORTS_GOLD_LINKER)
+endif()
+
+if (_COMPILER_SUPPORTS_GOLD_LINKER)
+  option(USE_GOLD_LINKER "Use ld.gold linker (decreases linking time if available)")
+  mark_as_advanced(USE_GOLD_LINKER)
+endif()
+
+if(USE_GOLD_LINKER AND _COMPILER_SUPPORTS_GOLD_LINKER)
+  set(CMAKE_EXE_LINKER_FLAGS "-fuse-ld=gold ${CMAKE_EXE_LINKER_FLAGS}")
+  set(CMAKE_MODULE_LINKER_FLAGS "-fuse-ld=gold ${CMAKE_MODULE_LINKER_FLAGS}")
+  set(CMAKE_SHARED_LINKER_FLAGS "-fuse-ld=gold ${CMAKE_SHARED_LINKER_FLAGS}")
+endif()
```

Comment:
```suggestion
if(UNIX AND NOT APPLE)
  message(STATUS "Checking if ld.gold linker is supported")

  cmake_push_check_state()
  set(CMAKE_REQUIRED_QUIET TRUE)
  set(CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS} -fuse-ld=gold")
  check_cxx_source_compiles("int main() {return 0;}" Slicer_COMPILER_SUPPORTS_GOLD_LINKER)
  cmake_pop_check_state()

  set(_status "no")
  if(Slicer_COMPILER_SUPPORTS_GOLD_LINKER)
    set(_status "yes")
  endif()
  message(STATUS "Checking if ld.gold linker is supported - ${_status}")
  
  if(Slicer_COMPILER_SUPPORTS_GOLD_LINKER)
    option(Slicer_USE_GOLD_LINKER "Use ld.gold linker (decreases linking time if available)")
    mark_as_advanced(Slicer_USE_GOLD_LINKER)
    mark_as_superbuild(Slicer_USE_GOLD_LINKER)
    
    if(Slicer_USE_GOLD_LINKER)
      set(CMAKE_EXE_LINKER_FLAGS "-fuse-ld=gold ${CMAKE_EXE_LINKER_FLAGS}"
        CACHE STRING "Flags used by the linker during all build types."
        FORCE)
      set(CMAKE_MODULE_LINKER_FLAGS "-fuse-ld=gold ${CMAKE_MODULE_LINKER_FLAGS}"
        CACHE STRING "Flags used by the linker during the creation of modules during during all build types."
        FORCE)
      set(CMAKE_SHARED_LINKER_FLAGS "-fuse-ld=gold ${CMAKE_SHARED_LINKER_FLAGS}"
        CACHE STRING "Flags used by the linker during the creation of shared libraries  during all build types."
        FORCE)
    endif()
  endif()
else()
  set(Slicer_COMPILER_SUPPORTS_GOLD_LINKER OFF)
endif()

mark_as_superbuild(
  VARS
    CMAKE_EXE_LINKER_FLAGS:STRING
    CMAKE_MODULE_LINKER_FLAGS:PATH
    CMAKE_SHARED_LINKER_FLAGS:STRING
  ALL_PROJECTS
  )
```

Summary:
* Rename `USE_GOLD_LINKER/_COMPILER_SUPPORTS_GOLD_LINKER` to `Slicer_USE_GOLD_LINKER/Slicer_COMPILER_SUPPORTS_GOLD_LINKER` (similarly to `Slicer_LINKER_NO_AS_NEEDED_FLAG_REQUIRED`)
* If defined, ensure `Slicer_USE_GOLD_LINKER` is propagated to inner build
* Revisit logic to match how `-Wl,--no-as-needed` is handled
* Perform check only on Linux, there is no gold linked on macOS
* Always display status message related to linker support (not only during initial config)
* Force `CMAKE_*_LINKER_FLAGS` values into the CMake cache, this will ensure that toggling the option at the top-level will work if done in an existing build

-----

:warning: To ensure that the call to `mark_as_superbuild` works, move the complete block after

```
#-----------------------------------------------------------------------------
set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_SOURCE_DIR}/CMake" ${CMAKE_MODULE_PATH})
set(SUPERBUILD_TOPLEVEL_PROJECT Slicer)
list(APPEND EXTERNAL_PROJECT_ADDITIONAL_DIRS ${CMAKE_CURRENT_SOURCE_DIR}/SuperBuild)
include(ExternalProject)
include(ExternalProjectDependency)
include(ExternalProjectDependencyForPython)
include(ExternalProjectGenerateProjectDescription)
```

-----

The use of `mark_as_superbuild()` with `ALL_PROJECTS` will ensure the `CMAKE_*_LINKER_FLAGS` variables are associated with all external projects (including the one used in the context of custom applications)


-----

Consider adding `Slicer_USE_GOLD_LINKER` and `Slicer_COMPILER_SUPPORTS_GOLD_LINKER` to `SlicerConfig`, and also update `UseSlicer` module, this ensure that extensions are also build using the relevant linker.

Reply:
Calling `mark_as_superbuild()` with `ALL_PROJECTS` to ensure `CMAKE_*_LINKER_FLAGS` variables are also propagated to external projects, will have to be done there as well.

Reply:
> Calling `mark_as_superbuild()` with `ALL_PROJECTS` to ensure `CMAKE_*_LINKER_FLAGS` variables are also propagated to external projects will have to be added in **CTK** as well.

@jcfr Maybe CTK should have a `CTK_USE_GOLD_LINKER` variable driving the same behaviour as in this PR?